atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/ThriftObjectSizeUtils.java

```diff
@@ -284,15 +287,47 @@ private static <T> long getBytesSize(T data, Function<T, ByteBuffer> toByteBuffe
         return byteBuffer.remaining();
     }
 
-    public static <T> long getCollectionSize(Collection<T> collection, Function<T, Long> sizeFunction) {
+    public static <T> long getCollectionSize(@Nullable Collection<T> collection, ToLongFunction<T> sizeFunction) {
         if (collection == null) {
             return getNullSize();
         }
 
-        long sum = 0;
-        for (T item : collection) {
-            sum += sizeFunction.apply(item);
+        // random access lists can be more efficiently accessed via List::get(int)
+        // as this avoids allocating iterator
+        if (isRandomAccessList(collection)) {
+            return getCollectionSize((List<T>) collection, sizeFunction);
+        }
+        return sumSizes(collection, sizeFunction);
+    }
+
+    private static <T> long getCollectionSize(@Nullable List<T> list, ToLongFunction<T> sizeFunction) {
+        if (list == null) {
+            return getNullSize();
+        }
+
+        // random access lists can be more efficiently accessed via List::get(int)
+        // as this avoids allocating iterator
+        if (isRandomAccessList(list)) {
+            long sum = 0L;
+            //noinspection ForLoopReplaceableByForEach -- performance sensitive
+            for (int i = 0; i < list.size(); i++) {
+                sum += sizeFunction.applyAsLong(list.get(i));
+            }
+            return sum;
+        }
+
+        return sumSizes(list, sizeFunction);
+    }
+
+    private static <T> long sumSizes(Collection<T> collection, ToLongFunction<T> sizeFunction) {
+        long sum = 0L;
+        for (T t : collection) {
+            sum += sizeFunction.applyAsLong(t);
         }
         return sum;
```

Comment:
I debated using `return collection.stream().mapToLong(sizeFunction).sum();` but opted to avoid stream overhead here